sled-agent/src/bootstrap/agent.rs

```diff
@@ -89,74 +125,84 @@ impl Agent {
-                        // "addr.to_string()" results in an IP address format
-                        // that is rejected when embedded into a URL.
-                        //
-                        // Instead, we merely use IP and port for the moment,
-                        // which loses the scope information. Longer-term, if we
-                        // use ULAs (Unique Local Addresses) the scope shouldn't
-                        // be a factor anyway.
-                        let addr_str = format!("[{}]:{}", addr.ip(), addr.port());
-                        info!(&self.log, "bootstrap: Connecting to {}", addr_str);
-                        BootstrapClient::new(
-                            &format!("http://{}", addr_str),
-                            self.log.new(o!(
-                                "Address" => addr_str,
-                            )),
+                        addr.set_port(trust_quorum::PORT);
+                        trust_quorum::Client::new(
+                            &self.log,
+                            share().verifier.clone(),
+                            addr,
                         )
                     })
                     .collect();
+
+                // TODO: Parallelize this and keep track of who's shares we've already retrieved and
+                // don't resend. See https://github.com/oxidecomputer/omicron/issues/514
+                let mut shares = vec![share().share.clone()];
                 for agent in &other_agents {
-                    agent
-                        .api_request_share(&bootstrap_types::ShareRequest {
-                            identity: vec![],
-                        })
-                        .await
+                    let share = agent.get_share().await
                         .map_err(|e| {
-                            info!(&self.log, "Bootstrap: Failed to share request with peer: {:?}", e);
-                            BackoffError::Transient(BootstrapError::Api(e))
+			    info!(&self.log, "Bootstrap: failed to retreive share from peer: {:?}", e);
+                            BackoffError::Transient(e)
                         })?;
-                        info!(&self.log, "Bootstrap: Shared request with peer");
+                    info!(
+                        &self.log,
+                        "Bootstrap: retreived share from peer: {}",
+                        agent.addr()
+                    );
+                    shares.push(share);
                 }
-                Ok(())
+                let rack_secret = RackSecret::combine_shares(
+                    share().threshold,
+                    share().total_shares,
+                    &shares,
+                )
+                .map_err(|e| {
+                    warn!(
+                        &self.log,
+                        "Bootstrap: failed to construct rack secret: {:?}", e
+                    );
+                    // TODO: We probably need to actually write an error
```

Comment:
(No action needed)

Thanks for filing these bugs! Makes it way easier to reason about real-but-nasty failure conditions we should eventually be handling.